src/Castle.Core/DynamicProxy/Contributors/ClassProxySerializableContributor.cs

```diff
@@ -29,29 +32,55 @@ namespace Castle.DynamicProxy.Contributors
 
 	internal class ClassProxySerializableContributor : SerializableContributor
 	{
-		private readonly bool delegateToBaseGetObjectData;
-		private readonly bool implementISerializable;
+		private bool delegateToBaseGetObjectData;
 		private ConstructorInfo serializationConstructor;
 		private readonly IList<FieldReference> serializedFields = new List<FieldReference>();
 
-		public ClassProxySerializableContributor(Type targetType, IList<MethodInfo> methodsToSkip, Type[] interfaces,
-		                                     string typeId)
+		public ClassProxySerializableContributor(Type targetType, Type[] interfaces, string typeId)
 			: base(targetType, interfaces, typeId)
 		{
-			if (targetType.IsSerializable)
+			Debug.Assert(targetType.IsSerializable, "This contributor is intended for serializable types only.");
+		}
+
+		public override void CollectElementsToProxy(IProxyGenerationHook hook, MetaType model)
+		{
+			delegateToBaseGetObjectData = VerifyIfBaseImplementsGetObjectData(targetType, model, out var getObjectData);
+
+			// This contributor is going to add a `GetObjectData` method to the proxy type.
+			// If a method with the same name and signature exists in the proxied class type,
+			// and another contributor has decided to proxy it, we need to tell it not to.
+			// Otherwise, we'll end up with two implementations!
+
+			if (getObjectData == null)
 			{
-				implementISerializable = true;
-				delegateToBaseGetObjectData = VerifyIfBaseImplementsGetObjectData(targetType, methodsToSkip);
+				// `VerifyIfBaseImplementsGetObjectData` only searches for `GetObjectData`
+				// in the implementation map for `ISerializable`. In the best case, it was
+				// already found there. If not, we need to look again, since *any* method
+				// with the same signature is a problem.
+
+				var getObjectDataMethod = targetType.GetMethod(
+					"GetObjectData",
+					BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance,
+					null,
+					new[] { typeof(SerializationInfo), typeof(StreamingContext) },
+					null);
+
+				if (getObjectDataMethod != null)
+				{
+					getObjectData = model.Methods.FirstOrDefault(m => m.Method == getObjectDataMethod);
```

Comment:
I know we've had a few cases of people reporting performance problems when they've got classes with hundreds or thousands of methods, yep some crazy people, but it'd be pretty quick to add a `MethodInfo` index something like this. At minimum you'd get rid of the Linq extension method usage which hurts performance.

```csharp
internal class MetaType
{
    private readonly Dictionary<MethodInfo, MetaMethod> methodsIndex = new Dictionary<MethodInfo, MetaMethod>();

    public MetaMethod FindMethod(MethodInfo info)
    {
        methodsIndex.TryGetValue(info, out var MetaMethod method);
        return method;
    }
}
```

Reply:
Done. We could go even further and merge `methodsIndex` with `methods` (which is a `TypeElementCollection`). Noone actually ever queries `MetaType.Methods`, it's only there to prevent method name conflicts across all contributors. So the type of collection used internally doesn't really matter.

However I suggest we do that separately, there are a few other places in DP where we could optimize the type of collection used.

Reply:
Wasn't line 70 meant to use `FindMethod` too?

Reply:
Uhh, yes, but then confusion struck. 🤦 Sorry about that. Fixed.